astropy/table/index.py

```diff
@@ -0,0 +1,781 @@
+shift_left(row) -> None : decrement row numbers after row
+shift_right(row) -> None : increase row numbers >= row
+find(key) -> list : list of rows corresponding to key
+range(lower, upper, bounds) -> list : rows in self[k] where k is between
+                               lower and upper (<= or < based on bounds)
+sort() -> list of rows in sorted order (by key)
+replace_rows(row_map) -> None : replace row numbers based on slice
+items() -> list of tuples of the form (key, data)
+
+Notes
+-----
+    When a Table is initialized from another Table, indices are
+    (deep) copied and their columns are set to the columns of the new Table.
+
+    Column creation:
+    Column(c) -> deep copy of indices
+    c[[1, 2]] -> deep copy and reordering of indices
+    c[1:2] -> reference
+    array.view(Column) -> no indices
+'''
+
+class QueryError(ValueError):
+    '''
+    Indicates that a given index cannot handle the supplied query.
+    '''
+    pass
+
+class Index(object):
+    '''
+    The Index class makes it possible to maintain indices
+    on columns of a Table, so that column values can be queried
+    quickly and efficiently. Column values are stored in lexicographic
+    sorted order, which allows for binary searching in O(log n).
+
+    Parameters
+    ----------
+    columns : list or None
+        List of columns on which to create an index. If None,
+        create an empty index for purposes of deep copying.
+    engine : type or None
+        Indexing engine class to use, from among SortedArray, BST,
+        FastBST, and FastRBT. If the supplied argument is None (by
+        default), use SortedArray.
+    data : SortedArray, BST, FastBST, FastRBT, or None
+        Engine data to copy
+    '''
+    def __new__(cls, *args, **kwargs):
+        self = super(Index, cls).__new__(cls)
+        self.__init__(*args, **kwargs)
+        return SlicedIndex(self, slice(0, 0, None), original=True)
+
+    def __init__(self, columns, engine=None, data=None):
+        from .table import Table
+
+        if data is not None: # create from data
+            self.engine = data.__class__
+            self.data = data
+            self.columns = columns
+            return
+
+        # by default, use SortedArray
+        self.engine = engine or SortedArray
+
+        if columns is None: # this creates a special exception for deep copying
+            columns = []
+            data = []
+            row_index = []
+        elif len(columns) == 0:
+            raise ValueError("Cannot create index without at least one column")
+        else:
+            num_rows = len(columns[0])
+
+            # replace Time columns with approximate form and remainder
+            new_columns = []
+            for col in columns:
+                if isinstance(col, Time):
+                    new_columns.append(col.jd)
+                    remainder = col - col.__class__(col.jd, format='jd')
+                    new_columns.append(remainder.jd)
```

Comment:
I don't think worth bothering with right now, but it might be nice if an interface were added to mixin columns that allow them to provide a view or transformation of the values in the column for use in lexical sorting.